### DISCUSSION TITLE
Creating Gordon Surfaces

### DISCUSSION THREAD

**Discussion by jsbroks** [gumyr/build123d#830] [Q&A] [answered]
I'm wondering what is the best way to create a Gordon Surface?

Here is what I have so far but as you would expect I getting `Error building a non-planer face with the provided exterior`.

```python
import csv
import numpy as np
from build123d import *


def read_heightmap(filename: str):
    with open(filename, 'r') as f:
        reader = csv.reader(f)
        heightmap = []
        for row in reader:
            heightmap.append([float(val) for val in row])
    return np.array(heightmap)


heightmap = read_heightmap('heightmap.csv')

points = []
for x in range(heightmap.shape[0]):
    for y in range(heightmap.shape[1]):
        points.append(Vector(x, y, heightmap[x,y]))


x_splines = []
y_splines = []


for x in range(heightmap.shape[0]):
    points = [(x, y, heightmap[x,y]) for y in range(heightmap.shape[1])]
    x_splines.append(Spline(points))


for y in range(heightmap.shape[1]):
    points = [(x, y, heightmap[x,y]) for x in range(heightmap.shape[0])]
    y_splines.append(Spline(points))

boundary_wire = Wire.make_wire([x_splines[0], y_splines[0], x_splines[-1], y_splines[-1]])

show_object(Face.make_surface(boundary_wire))
```

**Comment by gumyr** [comment]
The [make_surface_from_array_of_points](https://build123d.readthedocs.io/en/latest/direct_api_reference.html#topology.Face.make_surface_from_array_of_points) method might be what you want here.  

A Gordon Surface isn't currently supported by build123d; however, to create one it looks like GeomFill_AppSurf might be the correct OpenCascade class.  Here is a rough (AI generated) template how this might work:
```py
#include <GeomFill_AppSurf.hxx>
#include <Geom_BSplineCurve.hxx>
#include <GeomFill_SimpleBound.hxx>
#include <Geom_Surface.hxx>
#include <TopoDS_Face.hxx>
#include <BRepBuilderAPI_MakeFace.hxx>

void CreateGordonSurface() {
    // Example boundary curves (replace with your actual curves)
    Handle(Geom_BSplineCurve) curve1 = ...; // U direction curve 1
    Handle(Geom_BSplineCurve) curve2 = ...; // U direction curve 2
    Handle(Geom_BSplineCurve) curve3 = ...; // V direction curve 1
    Handle(Geom_BSplineCurve) curve4 = ...; // V direction curve 2

    // Create bounds for U and V directions
    GeomFill_SimpleBound uBound1(curve1, GeomAbs_C0);
    GeomFill_SimpleBound uBound2(curve2, GeomAbs_C0);
    GeomFill_SimpleBound vBound1(curve3, GeomAbs_C0);
    GeomFill_SimpleBound vBound2(curve4, GeomAbs_C0);

    // Create the Gordon surface
    GeomFill_AppSurf gordonSurface;
    gordonSurface.Init(uBound1, uBound2, Standard_True, Standard_True);
    gordonSurface.Perform(vBound1, vBound2, Standard_True, Standard_True);

    // Retrieve the resulting surface
    Handle(Geom_Surface) surface = gordonSurface.Surface();

    // Optional: Convert to a TopoDS_Face
    TopoDS_Face face = BRepBuilderAPI_MakeFace(surface, Precision::Confusion());

    // Use the face in your model
}
```

**Comment by mnesarco** [comment, marked as answer]
TiGL[1] has an OpenCascade based Gordon surface builder that is very good (c++). Christophe Grellier (@tomate44) ported it to FreeCAD's Curves WB[2] long time ago in **Python**. It can be used as a reference to port it to build123d. I am in the process of reorganizing the implementation as a standalone python module, but it is WIP and i can't make compromises on that [3]

[1] https://dlr-sc.github.io/tigl/
[2] https://github.com/tomate44/CurvesWB/blob/master/freecad/Curves/gordon.py
[3] https://github.com/mnesarco/MarzWorkbench/blob/master/freecad/marz/curves/gordon.py

**Reply by gumyr** [reply]
Excellent, thank you!

**Reply by tomate44** [reply]
@rainman110 made a standalone version of the TiGL Gordon Surface : 
https://github.com/rainman110/occ_gordon

**Reply by jsbroks** [reply]
TiGL got me what I wanted! ty!

**Reply by rainman110** [reply]
@mnesarco as @tomate44 pointed out, I created a standalone version of the tigl code, including python bindings created with swig. So you can use it together with pythonocc. I though still need to create conda packages .

**Reply by gumyr** [reply]
This is some truly excellent work - I'd love to have this available to build123d users.  

@jdegenstein, could we make the packaging work with pythonocc?  I would have thought that it would need to be OCP based and available on pypi for build123d to import it as a Python package.

**Reply by jdegenstein** [reply]
@gumyr I doubt we can make it work with pythonocc, but it may be possible to recompile against OCP somehow. I am definitely interested in what @rainman110 has done here, it is potentially a huge feature addition for build123d.

**Reply by rainman110** [reply]
@jdegenstein + @gumyr I am still pretty flexible. In fact, the library offers only just one api function (interpolate_curve_network), so wrapping it to python is straight forward. I though have no clue, how build123d interfaces occt. But i support, occ_gordon can be interfaced in the same manner.

**Reply by gumyr** [reply]
We use the CadQuery OCP based Python wrapper of OCCT instead of pythonOCC but I'm not sure that would matter with just a single Python API. Making a pypi package would be the bulk of the work I expect but I don't have experience in doing this.

**Reply by jdegenstein** [reply]
I think when occ_gordon is compiled (from C++) it needs to link against the compiled OCCT libs too and I doubt that OCP compiled libs can be substituted for the pythonOCC ones. That being said it shouldn't be impossible to accomplish.

**Reply by rainman110** [reply]
I suppose we just need to wrap occ_gordon with pybind11, which ocp seems to be also using (via bindgen). Then it could be included into ocp.

**Reply by rainman110** [reply]
If it helps. I also plan to release occ_gordon as a header file only library. This should make it much easier to integrate into other tools like ocp.

**Reply by mnesarco** [reply]
> If it helps. I also plan to release occ_gordon as a header file only library. This should make it much easier to integrate into other tools like ocp.

You can contribute it directly to OpenCascade to a wider audience. OpenCascade is now on github and they are actively accepting code contributions.

**Reply by gumyr** [reply]
Maybe this is a silly idea, but at first glance it looks like the entire Gordon Surface code could be converted to Python without too much trouble. This would avoid all of the difficulty of C++ with the cost of being somewhat slower.  How much time does it take to create the engine cowling with the current implementation?

**Reply by mnesarco** [reply]
> Maybe this is a silly idea, but at first glance it looks like the entire Gordon Surface code could be converted to Python without too much trouble. This would avoid all of the difficulty of C++ with the cost of being somewhat slower.  How much time does it take to create the engine cowling with the current implementation?

The @tomate44 port is pure python. I repackaged it into a single python file and removed unused code. It can be used as a starting point.

**Reply by gumyr** [reply]
Ah - if it's already pure Python then we only have to convert from pythonOCC to OCP which isn't a big deal.  It could then be added as a build123d method and fully integrated.

**Reply by mnesarco** [reply]
> Ah - if it's already pure Python then we only have to convert from pythonOCC to OCP which isn't a big deal.  It could then be added as a build123d method and fully integrated. 

It does not even use pythonocct. I think it would be easy to port to pure ocp.
